go/mysql/conn.go

```diff
@@ -136,8 +135,8 @@ type Conn struct {
 	ClientData interface{}
 
 	// Packet encoding variables.
-	reader   *bufio.Reader
-	writer   *bufio.Writer
+	reader   bufioReader
+	writer   bufioWriter
 	sequence uint8
 
```

Comment:
github won't let me comment on other lines, so here i go.

as mentioned, since most connections sit waiting for read, "pooling" doesn't help much.
however, we can note that every read must read the 4 byte length header.

this leaves us a few options:
* write a read header function, which does a direct read from the base connection (and loops until it gets its 4 bytes)
* initialize all conns with a 4 byte buffered reader.

only once we've read the header do we then use the pooled buffered reader.

in both cases i think factoring out the header read is useful so then we don't risk someone messing up the header read by using the bufio.

@sougou thoughts? you'll notice option 2 is just the "1 byte buffer" option repackaged slightly to match the data we already need to read ;)

Reply:
In spite of the benchmarks, I don't think the pooling will give us anything in real life. This is because most connections will generally be idle, and during that time, they will be possessing a buffer waiting on a read.

If we went with the 4-byte approach, we'll need to perform some real-life benchmarks to evaluate the impact of the extra syscall. If we find that it's worth it, then we should not use bufio at all:
All waiting reads can fetch into a 4-byte buffer, and then we can directly read the rest of the data into the target slice. There is no benefit to using bufio in this case.

Another option is to reduce the size of bufio to 4K, which is what go recommends. With the removal of the other two buffers, the overhead of a connection would drop to 4K, which can be more easily accommodated.

We should just look at dropping connBuffSize and use go's default anyways.